### DISCUSSION TITLE
How can to use "SPI oled x2 example" with I2C oled? (GND/VCC/SCL/SDA).

### DISCUSSION THREAD

**Discussion by WeaponX15** [tttapa/Control-Surface#580] [Q&A]
Hi Peterp, How I can to use "SPI oled x2 example"(MCU-OLED-SSD1306-X2),  with I2C oled? (GND/VCC/SCL/SDA).

Thanks.

**Comment by tttapa** [comment]
Where exactly are you stuck?  
There is no difference in the Control-Surface specific code, switching between I2C and SPI is done entirely using the Adafruit SSD1306 library.

**Reply by WeaponX15** [reply]
I want view trackname or track info in oled display, and just shows two "semicircles and one line" (i use reaper), thanks.

**Reply by tttapa** [reply]
This sounds like a MIDI problem rather than a display problem.

Receiving track names is handled by the [`MCU::LCD<>`](https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d2e/classMCU_1_1LCD.html) class.  
Did you map your Arduino as a Mackie Control Universal unit in Reaper? Have you inspected the MIDI messages you're receiving (e.g. using the second sketch in the [routing between MIDI interfaces section in the MIDI tutorial](https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/df7/midi-tutorial.html#midi_md-routing-between-interfaces))?

**Reply by WeaponX15** [reply]
The function of displaying track name or data etc on the screen, does it only work with the MCU protocol?

**Reply by WeaponX15** [reply]
Connect pins 20-21 to sda / scl of the oled. I can move the pan knob 1, but when I change the bank (channel 2) and the pan of track 1 continues to move and not track 9. And I still don't have any image in the oled, I just settle for the name of the track, only that.

The other question I have with the sketch of the MCU protocol, how can I assign the control mapping (the first knob is automatically assigned to pan channel 1), how can I assign the volume fader track 1?

 Very thanks.

**Reply by WeaponX15** [reply]
upgrade; Modify sketch, set and enable MCU in Reaper, managed to show pan and number track, and move pan knob in oled, but couldn't get track name (check MCU :: LCD class and add). The problem is that everything was destabilized in Reaper, the controls became intermittent and inaccurate and reaper came back very slow, I don't know what it can be, I eliminate the MCU in reaper and everything returns to normal.

**Reply by tttapa** [reply]
> The function of displaying track name or data etc on the screen, does it only work with the MCU protocol?

AFAIK, the MCU protocol is the only “standard” way to do this. If your DAW supports other ways or formats to communicate the track names, Control Surface provides all the tools you need, but because there's no standard protocol that I'm aware of, you'll have to do the parsing of the message format yourself. You can see how the `MCU::LCD` class does this for the MCU format, other formats will be very similar so you can reuse most code.

> Connect pins 20-21 to sda / scl of the oled.

The I²C hardware peripheral of the Arduino's microcontroller is connected to specific pins, you have to use these pins (unless your board supports changing the I²C pins in software, the Mega does not).

The connection and usage of the OLED display is handled entirely by the Adafruit SSD1306 library, you'll have to consult its documentation and examples, e.g. https://github.com/adafruit/Adafruit_SSD1306/blob/master/examples/ssd1306_128x64_i2c/ssd1306_128x64_i2c.ino.  
In your code, it should look like this:
```cpp
constexpr uint8_t SCREEN_WIDTH = 128;
constexpr uint8_t SCREEN_HEIGHT = 64;
constexpr int8_t OLED_reset = -1; // Use the external RC circuit for reset

// Instantiate the displays
Adafruit_SSD1306 ssd1306Display {
  SCREEN_WIDTH, SCREEN_HEIGHT, &Wire, OLED_reset,
};
```
> I can move the pan knob 1, but when I change the bank (channel 2) and the pan of track 1 continues to move and not track 9.

Which knob? The example I'm looking at doesn't define any knobs. Please post your (complete) code.  
What type of banks are you using? Control Surface banks (`Bank<N>`) or MCU banks?

> The other question I have with the sketch of the MCU protocol, how can I assign the control mapping (the first knob is automatically assigned to pan channel 1), how can I assign the volume fader track 1?

This is one of the downsides of the MCU protocol, it's not flexible, you cannot change the mapping yourself. The volume control for MCU uses pitch bend controls on channels 1-8, see [MCU Pitch Bend controls
](https://tttapa.github.io/Control-Surface-doc/Doxygen/da/dc1/group__MCU__PB.html).  
You could use a `PBPotentiometer` or a `PBAbsoluteEncoder`.

Alternatively, don't use the MCU protocol. IIRC, you can define custom control surfaces in Reaper (not sure if you can get it to send the track names, though). But that's a question for the Reaper forum, I can't really help you with that. Once you get Reaper to send it, I can show you how to receive it using Control Surface, but I'm not familiar enough with Reaper.

> but couldn't get track name

Try removing this check:
https://github.com/tttapa/Control-Surface/blob/feee931d44d5f20fb7913b171895316e5819549d/src/Display/MCU/LCDDisplay.hpp#L73-L75

> The problem is that everything was destabilized in Reaper, the controls became intermittent and inaccurate and reaper came back very slow, I don't know what it can be, I eliminate the MCU in reaper and everything returns to normal.

The Arduino Mega only supports MIDI over Serial, this is extremely slow compared to USB communication. The MCU protocol exchanges large amounts of data, sending all of this over a serial link takes a significant amount of time. Reaper has to wait for all of this data to be transmitted before it can continue. This is arguably poor design on Reaper's end, this blocking MIDI communication shouldn't happen in the UI thread, but the key problem here is the slow serial link to the Arduino.  
I've only ever used the MCU protocol on USB-capable microcontrollers like the Teensy.

**Comment by WeaponX15** [comment]
The Oled / Adafruit code that you indicate is correct, it looks as you indicate it, in fact it already communicates correctly, except for the name of the track. As for this point I think that here is the error, I just added the MCUL class :: LCD, but not the complete code that you indicate (LCDDisplay.hpp) I tried to add it (copy / paste) but it starts throwing compilation errors, sorry PeterP but at the moment my knowledge in C ++ is limited, I have learned a lot and I It allows us to understand and modify some codes, but as for obtaining values ​​in Display hardware I still understand very little, how should LCDDisplay.hpp be inserted in the sketch?

Another thing, I have Hairless and loopMidi running, then I run Reaper and start monitoring a large amount of outgoing data in loopMidi and Hairless (I obviously imagine that it is the MCU protocol communicating the state changes of the controls) and loopmidi crashes and throws error (muted). Hairless constantly receives the message "Midi In: Ch 1: Note 25 on velocity 127/0".

I have eliminated from the code values ​​or messages that the MCU protocol sends to arduino to reduce and lighten the data load, but loopmidi keeps blocking and reaper becomes excessively slow, I think it may be that my code / sketch is little optimized, or that it simply as you point out, Mega arduino does not have the ability to handle it.

Finally, do you think that if I use 8 oled to show only the names of the tracks without any other information it can be tolerated by mega arduino? I have a Leonardo but the code exceeds the Leonardo's capacity and I just can't load the sketch.

Thank you again Peter for your patience and willingness.

**Reply by tttapa** [reply]
> I just added the MCUL class :: LCD, but not the complete code that you indicate (LCDDisplay.hpp) I tried to add it (copy / paste) but it starts throwing compilation errors

Don't add it to your sketch, comment out the lines I mentioned in the library file in your Arduino library installation folder.

> Hairless constantly receives the message "Midi In: Ch 1: Note 25 on velocity 127/0".

Receives from where? The Arduino or Reaper? Note 25 is the “select track 2” message in the MCU protocol.

> I have eliminated from the code values ​​or messages that the MCU protocol sends to arduino to reduce and lighten the data load

How?

> reaper becomes excessively slow, I think it may be that my code / sketch is little optimized

This is not caused by your sketch, it is caused by the slow serial connection you're using.

> Finally, do you think that if I use 8 oled to show only the names of the tracks without any other information it can be tolerated by mega arduino? 

The problem you're seeing is not caused by what your Arduino code is doing, it is caused by the large amount of data Reaper is trying to send over a slow link. Either increase the speed of the link, or reduce the amount of data Reaper tries to send.

Either way, you cannot use 8 OLED displays on an Arduino Mega using the Adafruit library, because it allocates a 1KiB buffer for each display, you only have 8KiB of RAM to work with. Control Surface supports sharing one buffer for all displays, but the Adafruit SSD1306 library does not. I have a fork of the Adafruit library that does support it.

**Comment by WeaponX15** [comment]
I think I will upgrade to a Teensy 4 as you recommend, the project with Mega I will complement it with OSC, the problem is that after the errors with LoopMidi and Reaper, I have lost total communication with Reaper (Leonardo and Mega), something got corrupted in Windows / Reaper (I tried on another pc with reaper and everything was correct) reinstall everything (reaper, loopmidi) and reaper continues without receiving midi messages and I cannot map anything, I will solve it, I will acquire a Teensy and return to the attack.
